ratbagctl rewrite and test suite

bentiss opened the pull request:
Well, there are still a few FIXMEs here and there, but better sending this one out before it gets too much of a beast.

The idea is to instantiate a driver-test libratbag element, and playing with unit tests to see if we messed up the parsing of rabtagctl.

This gives a good idea if ratbagctl/ratbagd works as expected.

The current test run takes roughly 21 seconds on my machine, most of the time is spent by forking, initializing the bus, and settings the devices, etc...

I have a local version that creates only one dbus connection and directly calls the parsing/function from ratbagctl. This version runs in less than a second but fails because ratbagctl python binding doesn't listen for signals and can't update its internal state. I tried playing with the GLib.MainLoop, without success.

whot commented on tools/ratbagctl.test.in:
```diff
@@ -0,0 +1,407 @@
+    global ratbagd_process
+    # first copy the policy for the ratbagd daemon to be allowed to run
+    shutil.copy(os.path.join(@MESON_BUILD_ROOT@, 'org.freedesktop.ratbag_devel1.conf'),
+                '/etc/dbus-1/system.d/')
```
shutil takes a file as second arg as well, I really think we should have a const here for the path that we pass into os.unlink() below as well.

whot commented on tools/ratbagctl.in:
```diff
@@ -1540,21 +1540,28 @@ class RatbagParserRoot(object):
+            print("No devices available.")
+
+        return r
+
```
just add an extra spare line here, I'm planning to fix ratbagctl for flake8, this will complain. Might as well fix it now already

bentiss replied:
Well, given the amount of flake8 errors this series introduce, I might as well add such a check before messing up too much in the python code.

whot commented on tools/ratbagctl.in:
```diff
@@ -1540,21 +1540,28 @@ class RatbagParserRoot(object):
+        return None
+    else:
+        if not r.devices:
+            print("No devices available.")
```
bit of a strange choice to use `else:` here, imo. Why not do
```
try:
   r = Ratbagd()
   if not r.devices:
       print()
   return r
except RatbagdDBusUnavailable:
   ...
````

bentiss replied:
https://docs.python.org/3/tutorial/errors.html
> The use of the `else` clause is better than adding additional code to the `try` clause because it avoids accidentally catching an exception that wasn’t raised by the code being protected by the `try ... except` statement.

I have been given example of these by the python core developer who helped me on the argparse mess. And I must say that I agree with the above statement.

bentiss commented on ratbagd/ratbagd.c:
```diff
@@ -181,6 +181,148 @@ static int ratbagd_get_themes(sd_bus *bus,
+	r = ratbagd_device_new(&ratbagd_test_device, ctx, "test_device", device);
+
+	/* the ratbagd_device takes its own reference, drop ours */
+	ratbag_device_unref(device);
```
Well, no. And actually, the error path is wrong in ratbagd_device_new().

Theorically, ratbagd_device_new() calls ratbag_device_ref() on the ratbag device and should unref it in case of an error. We do not want to keep the reference around anyway, so we should unref our own ratbag_device here so that if r is negative or null, both cases are fine.

The bug is more likely in ratbagd_device_new() which doesn't unref the device in case of a failure (though the failures cases are unlikely to happen on a normal execution path).

whot replied:
I'd say that's a bug in ratbagd_device_new, yes. Bit strange behaviour to call new and still have to cleanup afterwards.

bentiss commented:
Done. I amended according to your comments and added a .flake8 file (and fixed warnings).

flake8 circleci is still not triggered and ratbagd_device_new() still not fixed
